src/hardware/adc_internal.rs

```diff
@@ -0,0 +1,133 @@
+use super::hal::{
+    adc,
+    gpio::{gpioa::*, gpiob::*, gpioc::*, gpiof::*, Analog},
+    hal::blocking::delay::DelayUs,
+    prelude::*,
+    rcc::{rec, CoreClocks},
+    stm32::{ADC1, ADC2, ADC3},
+};
+
+pub enum TecChannel {
+    Zero,
+    One,
+    Two,
+    Three,
+}
+
+pub enum Supply {
+    P5v,
+    P12v,
+    P3v3,
+    I12v,
+}
+
+pub struct AdcInternal {
+    adc1: adc::Adc<ADC1, adc::Enabled>,
+    adc3: adc::Adc<ADC3, adc::Enabled>,
+    p5v: PC0<Analog>,
+    p12v: PC2<Analog>,
+    p3v3: PF7<Analog>,
+    i12v: PF8<Analog>,
+    tecu: TecUPins,
+    teci: TecIPins,
+}
+
+const V_REF: f32 = 3.0; // ADC reference voltage
+const DIV_P12V: f32 = 1.6 / (1.6 + 6.8); // Resistor divider 12V rail
+const DIV_P5V: f32 = 6.8 / (10.0 + 6.8); // Resistor divider 5V rail
+const DIV_P3V3: f32 = 6.8 / (1.6 + 6.8); // Resistor divider 3V3 rail
+const CONF_I12V: f32 = 0.005 * (10000.0 / 100.0); // 12V current measurement resistor configuration for LT6106
+
+pub type TecUPins = (PC3<Analog>, PA0<Analog>, PA3<Analog>, PA4<Analog>);
+pub type TecIPins = (PA5<Analog>, PA6<Analog>, PB0<Analog>, PB1<Analog>);
+
+impl AdcInternal {
+    #![allow(clippy::too_many_arguments)]
+    pub fn new(
+        delay: &mut impl DelayUs<u8>,
+        clocks: &CoreClocks,
+        adc12_rcc: rec::Adc12,
+        adc3_rcc: rec::Adc3,
+        adc1: ADC1,
+        adc2: ADC2,
+        adc3: ADC3,
+        p5v: PC0<Analog>,
+        p12v: PC2<Analog>,
+        p3v3: PF7<Analog>,
+        i12v: PF8<Analog>,
+        tecu: TecUPins,
+        teci: TecIPins,
+    ) -> Self {
+        // Setup ADC1 and ADC2
+        let (adc1, _) = adc::adc12(adc1, adc2, delay, adc12_rcc, clocks);
+        let adc3 = adc::Adc::adc3(adc3, delay, adc3_rcc, clocks);
+
+        let mut adc1 = adc1.enable();
+        adc1.set_resolution(adc::Resolution::SIXTEENBIT);
+
+        let mut adc3 = adc3.enable();
+        adc3.set_resolution(adc::Resolution::SIXTEENBIT);
+
+        AdcInternal {
+            adc1,
+            adc3,
+            p5v,
+            p12v,
+            p3v3,
+            i12v,
+            tecu,
+            teci,
+        }
+    }
+
+    pub fn read_tecu(&mut self, tecu: TecChannel) -> u32 {
```

Comment:
These should also get conversion methods to return SI units.

Reply:
I'll implement this in another step where the outputs can actually be configured aka when dac_pwm is merged.